Testing, suddenly a test module I have that uses GetBitmap() on wxMenuItem without any args is now giving

```
        const wxBitmap& bitmap = referenceItem->GetBitmap();
```
E:\kicad\kicad-wx-test\src\wxtest1\wxtest1.cpp(93): error C2668: 'wxMenuItem::GetBitmap': ambiguous call to overloaded function
  E:\vcpkg2\installed\x64-windows\include\wx/msw/menuitem.h(99): note: could be 'wxBitmap wxMenuItem::GetBitmap(bool) const'
  E:\vcpkg2\installed\x64-windows\include\wx/msw/menuitem.h(86): note: or       'wxBitmap wxMenuItem::GetBitmap(void) const'
  E:\kicad\kicad-wx-test\src\wxtest1\wxtest1.cpp(93): note: while trying to match the argument list '()'
```

--- a/include/wx/gtk/menuitem.h
+++ b/include/wx/gtk/menuitem.h
@@ -29,8 +29,6 @@ public:
     virtual void Enable( bool enable = true ) wxOVERRIDE;
     virtual void Check( bool check = true ) wxOVERRIDE;
     virtual bool IsChecked() const wxOVERRIDE;
-    virtual void SetBitmap(const wxBitmapBundle& bitmap);
-    virtual wxBitmap GetBitmap() const;
     void SetupBitmaps(wxWindow *win);
 
 #if wxUSE_ACCEL
@@ -60,7 +58,6 @@ public:
 #endif
 
 private:
-    wxBitmapBundle m_bitmap; // Bitmap for menuitem, if any
     GtkWidget *m_menuItem;  // GtkMenuItem
 
     wxDECLARE_DYNAMIC_CLASS(wxMenuItem);

--- a/include/wx/gtk1/menuitem.h
+++ b/include/wx/gtk1/menuitem.h
@@ -32,8 +32,6 @@ public:
     virtual void Enable( bool enable = TRUE );
     virtual void Check( bool check = TRUE );
     virtual bool IsChecked() const;
-    virtual void SetBitmap(const wxBitmapBundle& bitmap) { m_bitmap = bitmap; }
-    virtual wxBitmap GetBitmap() const;
 
 #if wxUSE_ACCEL
     virtual wxAcceleratorEntry *GetAccel() const;
@@ -65,7 +63,6 @@ private:
     void DoSetText(const wxString& text);
 
     wxString  m_hotKey;
-    wxBitmapBundle  m_bitmap; // Bitmap for menuitem, if any
 
     GtkWidget *m_menuItem;  // GtkMenuItem
     GtkWidget* m_labelWidget; // Label widget

--- a/include/wx/menuitem.h
+++ b/include/wx/menuitem.h
@@ -115,6 +115,11 @@ public:
     void SetHelp(const wxString& str);
     const wxString& GetHelp() const { return m_help; }
 
+    // bitmap-related functions
+
+    virtual void SetBitmap(const wxBitmapBundle& bmp);
+    virtual wxBitmap GetBitmap() const;
+
 #if wxUSE_ACCEL
     // extract the accelerator from the given menu string, return NULL if none
     // found
@@ -175,6 +180,7 @@ protected:
                  *m_subMenu;        // our sub menu or NULL
     wxString      m_text,           // label of the item
                   m_help;           // the help string for the item
+    wxBitmapBundle m_bitmap;        // item bitmap, may be invalid
     wxItemKind    m_kind;           // separator/normal/check/radio item?
     bool          m_isChecked;      // is checked?
     bool          m_isEnabled;      // is enabled?

--- a/include/wx/menuitem.h
+++ b/include/wx/menuitem.h
@@ -118,6 +118,10 @@ public:
     // bitmap-related functions
 
     virtual void SetBitmap(const wxBitmapBundle& bmp);
+    wxBitmapBundle GetBitmapBundle() const { return m_bitmap; }
+
+    // This method only exists for compatibility, prefer using
+    // GetBitmapBundle() in the new code.
     virtual wxBitmap GetBitmap() const;
 
 #if wxUSE_ACCEL

--- a/include/wx/motif/menuitem.h
+++ b/include/wx/motif/menuitem.h
@@ -35,12 +35,6 @@ public:
     virtual void SetItemLabel(const wxString& label);
     virtual void Enable(bool enable = true);
     virtual void Check(bool check = true);
-    // included SetBitmap and GetBitmap as copied from the GTK include file
-    // I'm not sure if this works but it silences the linker in the
-    // menu sample.
-    //     JJ
-    virtual void SetBitmap(const wxBitmapBundle& bitmap) { m_bitmap = bitmap; }
-    virtual wxBitmap GetBitmap() const { return GetBitmapFromBundle(m_bitmap); }
 
     // implementation from now on
     void CreateItem (WXWidget menu, wxMenuBar * menuBar, wxMenu * topMenu,
@@ -59,7 +53,6 @@ private:
     WXWidget    m_buttonWidget;
     wxMenuBar*  m_menuBar;
     wxMenu*     m_topMenu;        // Top-level menu e.g. popup-menu
-    wxBitmapBundle  m_bitmap; // Bitmap for menuitem, if any
 
     wxDECLARE_DYNAMIC_CLASS(wxMenuItem);
 };

--- a/include/wx/msw/menuitem.h
+++ b/include/wx/msw/menuitem.h
@@ -78,14 +78,25 @@ public:
         DoSetBitmap(bmpUnchecked, false);
     }
 
-    void SetBitmap(const wxBitmapBundle& bmp, bool bChecked = true)
+    virtual void SetBitmap(const wxBitmapBundle& bmp) wxOVERRIDE
     {
-        DoSetBitmap(bmp, bChecked);
+        DoSetBitmap(bmp, true);
+    }
+
+    virtual wxBitmap GetBitmap() const wxOVERRIDE
+    {
+        return GetBitmap(true);
     }
 
     void SetupBitmaps();
 
-    wxBitmap GetBitmap(bool bChecked = true) const;
+    // The functions taking bChecked are wxMSW-specific.
+    void SetBitmap(const wxBitmapBundle& bmp, bool bChecked)
+    {
+        DoSetBitmap(bmp, bChecked);
+    }
+
+    wxBitmap GetBitmap(bool bChecked) const;
 
 #if wxUSE_OWNER_DRAWN
     void SetDisabledBitmap(const wxBitmapBundle& bmpDisabled)
@@ -139,8 +150,7 @@ private:
     int MSGetMenuItemPos() const;
 
     // item bitmaps
-    wxBitmapBundle m_bitmap,     // bitmap to put near the item
-                   m_bmpUnchecked;   // (checked is used also for 'uncheckable' items)
+    wxBitmapBundle m_bmpUnchecked;   // (used only for checkable items)
 #if wxUSE_OWNER_DRAWN
     wxBitmapBundle m_bmpDisabled;
 #endif // wxUSE_OWNER_DRAWN

--- a/include/wx/osx/menuitem.h
+++ b/include/wx/osx/menuitem.h
@@ -48,9 +48,6 @@ public:
     void RemoveHiddenItems();
 #endif // wxUSE_ACCEL
 
-    virtual void SetBitmap(const wxBitmapBundle& bitmap) ;
-    virtual wxBitmap GetBitmap() const;
-
 
     // Implementation only from now on.
 
@@ -62,8 +59,6 @@ public:
     wxMenuItemImpl* GetPeer() { return m_peer; }
 private:
     void UncheckRadio() ;
-
-    wxBitmapBundle m_bitmap; // Bitmap for menuitem, if any
 
     wxMenuItemImpl* m_peer;
 

--- a/include/wx/qt/menuitem.h
+++ b/include/wx/qt/menuitem.h
@@ -35,8 +35,7 @@ public:
     virtual void Check(bool check = true) wxOVERRIDE;
     virtual bool IsChecked() const wxOVERRIDE;
 
-    virtual void SetBitmap(const wxBitmapBundle& bitmap);
-    virtual wxBitmap GetBitmap() const { return GetBitmapFromBundle(m_bitmap); }
+    virtual void SetBitmap(const wxBitmapBundle& bitmap) wxOVERRIDE;
 
     virtual QAction *GetHandle() const;
 
@@ -44,7 +43,6 @@ public:
 private:
     // Qt is using an action instead of a menu item.
     wxQtAction *m_qtAction;
-    wxBitmapBundle m_bitmap;
 
     wxDECLARE_DYNAMIC_CLASS( wxMenuItem );
 };

--- a/include/wx/univ/menuitem.h
+++ b/include/wx/univ/menuitem.h
@@ -39,9 +39,11 @@ public:
     // hopefully
     void SetBitmaps(const wxBitmapBundle& bmpChecked,
                     const wxBitmapBundle& bmpUnchecked = wxBitmapBundle());
-    void SetBitmap(const wxBitmapBundle& bmp) { SetBitmaps(bmp); }
-    wxBitmap GetBitmap(bool checked = true) const
+    virtual void SetBitmap(const wxBitmapBundle& bmp) wxOVERRIDE { SetBitmaps(bmp); }
+    wxBitmap GetBitmap(bool checked) const
       { return GetBitmapFromBundle(checked ? m_bitmap : m_bmpUnchecked); }
+    virtual wxBitmap GetBitmap() const wxOVERRIDE
+      { return GetBitmap(true); }
 
     void SetDisabledBitmap( const wxBitmapBundle& bmpDisabled )
       { m_bmpDisabled = bmpDisabled; }
@@ -93,8 +95,7 @@ protected:
     void UpdateAccelInfo();
 
     // the bitmaps (may be invalid, then they're not used)
-    wxBitmapBundle m_bitmap,
-                   m_bmpUnchecked,
+    wxBitmapBundle m_bmpUnchecked,
                    m_bmpDisabled;
 
     // the positions of the first and last items of the radio group this item

--- a/interface/wx/menuitem.h
+++ b/interface/wx/menuitem.h
@@ -162,6 +162,9 @@ public:
 
     /**
         Returns the item bitmap.
+
+        This method exists only for compatibility, please use GetBitmapBundle()
+        in the new code.
     */
     wxBitmap GetBitmap() const;
 
@@ -171,6 +174,18 @@ public:
         This overload only exists in wxMSW, avoid using it in portable code.
     */
     wxBitmap GetBitmap(bool checked) const;
+
+    /**
+        Returns the bitmap bundle containing the bitmap used for this item.
+
+        The returned bundle is invalid, i.e. empty, if no bitmap is associated
+        with the item.
+
+        @see SetBitmap()
+
+        @since 3.2.0
+    */
+    wxBitmapBundle GetBitmapBundle() const;
 
     /**
         Returns the bitmap used for disabled items.

--- a/src/common/menucmn.cpp
+++ b/src/common/menucmn.cpp
@@ -344,6 +344,16 @@ wxString wxMenuItemBase::GetLabelFromText(const wxString& text)
 }
 #endif
 
+void wxMenuItemBase::SetBitmap(const wxBitmapBundle& bmp)
+{
+    m_bitmap = bmp;
+}
+
+wxBitmap wxMenuItemBase::GetBitmap() const
+{
+    return GetBitmapFromBundle(m_bitmap);
+}
+
 wxBitmap wxMenuItemBase::GetBitmapFromBundle(const wxBitmapBundle& bundle) const
 {
     wxBitmap bmp;

--- a/src/gtk/menu.cpp
+++ b/src/gtk/menu.cpp
@@ -756,21 +756,6 @@ void wxMenuItem::ClearExtraAccels()
 
 #endif // wxUSE_ACCEL
 
-void wxMenuItem::SetBitmap(const wxBitmapBundle& bitmap)
-{
-    if (m_kind == wxITEM_NORMAL)
-        m_bitmap = bitmap;
-    else
-    {
-        wxFAIL_MSG("only normal menu items can have bitmaps");
-    }
-}
-
-wxBitmap wxMenuItem::GetBitmap() const
-{
-    return GetBitmapFromBundle(m_bitmap);
-}
-
 void wxMenuItem::SetupBitmaps(wxWindow *win)
 {
 #ifndef __WXGTK4__

--- a/src/gtk1/menu.cpp
+++ b/src/gtk1/menu.cpp
@@ -870,11 +870,6 @@ wxAcceleratorEntry *wxMenuItem::GetAccel() const
 
 #endif // wxUSE_ACCEL
 
-wxBitmap wxMenuItem::GetBitmap() const
-{
-   return GetBitmapFromBundle(m_bitmap);
-}
-
 void wxMenuItem::Check( bool check )
 {
     wxCHECK_RET( m_menuItem, wxT("invalid menu item") );

--- a/src/msw/window.cpp
+++ b/src/msw/window.cpp
@@ -2368,8 +2368,6 @@ static void wxYieldForCommandsOnly()
 
 bool wxWindowMSW::DoPopupMenu(wxMenu *menu, int x, int y)
 {
-    menu->SetupBitmaps();
-
     wxPoint pt;
     if ( x == wxDefaultCoord && y == wxDefaultCoord )
     {
@@ -2450,7 +2448,16 @@ wxWindowMSW::DoSendMenuOpenCloseEvent(wxEventType evtType, wxMenu* menu)
 {
     wxMenuEvent event(evtType, menu && !menu->IsAttached() ? wxID_ANY : 0, menu);
 
-    return wxMenu::ProcessMenuEvent(menu, event, this);
+    const bool processed = wxMenu::ProcessMenuEvent(menu, event, this);
+
+    if ( menu && evtType == wxEVT_MENU_OPEN )
+    {
+        // Do this after possibly executing the user-defined wxEVT_MENU_OPEN
+        // handler which could have modified the bitmaps.
+        menu->SetupBitmaps();
+    }
+
+    return processed;
 }
 
 bool wxWindowMSW::HandleMenuPopup(wxEventType evtType, WXHMENU hMenu)

--- a/src/osx/menuitem_osx.cpp
+++ b/src/osx/menuitem_osx.cpp
@@ -70,16 +70,6 @@ wxMenuItem::~wxMenuItem()
 // change item state
 // -----------------
 
-void wxMenuItem::SetBitmap(const wxBitmapBundle& bitmap)
-{
-      m_bitmap = bitmap;
-}
-
-wxBitmap wxMenuItem::GetBitmap() const
-{
-    return GetBitmapFromBundle(m_bitmap);
-}
-
 void wxMenuItem::Enable(bool bDoEnable)
 {
     if (( m_isEnabled != bDoEnable

